cranelift-codegen/src/legalizer/split.rs

```diff
@@ -131,19 +131,27 @@ fn split_any(
 }
 
 pub fn split_ebb_params(func: &mut ir::Function, cfg: &ControlFlowGraph, ebb: Ebb) {
-    let mut repairs = Vec::new();
     let pos = &mut FuncCursor::new(func).at_top(ebb);
+    let ebb_params = pos.func.dfg.ebb_params(ebb);
+
+    // Add further splittable types here.
+    fn type_requires_splitting(ty: Type) -> bool {
+        ty == ir::types::I128
+    }
 
-    for (num, ebb_param) in pos
-        .func
-        .dfg
-        .ebb_params(ebb)
-        .to_vec()
-        .into_iter()
-        .enumerate()
+    // A shortcut.  If none of the param types require splitting, exit now.  This helps because
+    // the loop below necessarily has to copy the ebb params into a new vector, so it's better to
+    // avoid doing so when possible.
+    if !ebb_params
+        .iter()
+        .any(|ebb_param| type_requires_splitting(pos.func.dfg.value_type(*ebb_param)))
     {
-        let ty = pos.func.dfg.value_type(ebb_param);
-        if ty != ir::types::I128 {
+        return;
+    }
+
+    let mut repairs = Vec::new();
+    for (num, ebb_param) in ebb_params.to_vec().into_iter().enumerate() {
```

Comment:
Maybe use

```rust
let split_ebb_params = ebb_params.iter().enumerate().filter_map(|(i, ebb_param)| {
    if type_requires_splitting(pos.func.dfg.value_type(ebb_param)) {
        Some(i, ebb_param)
    } else {
        None
    }
}).collect::<Vec<(usize, Ebb)>>();
for (num, ebb_param) in split_ebb_params.into_iter() {
    split_ebb_param(pos, ebb, num, ebb_param, Opcode::Iconcat, &mut repairs);
}
```

instead. That would make the allocated amount proportional to the amount of ebb params which require splitting, instead of proportional to the total amount of ebb params. You should then be able to remove the `if` above.

Reply:
This would still allocate two Vec plus call the tail function with an empty vector, right? This PR optimizes for minimizing work, which seems even nicer...

Reply:
@bnjbvr, yes, I think that's correct.

Reply:
It seems collect for `Vec` doesnt actually allocate anything until the first item. `Vec::new` doesnt actually allocate anything.

Reply:
The filter_map way would work too, but it's slightly less explicit in what it does. In particular, there's still need for stack allocations for the two Vecs, and the spurious call at the function tail. I don't think these particularly matter much, but the initial early exit is nice in that it does minimal work in every case (esp. when there's no ebb params on the current block). No strong opinions here, so I suggest to let it as is.